c10/core/CPUAllocator.cpp

```diff
@@ -323,7 +324,7 @@ void ProfiledCPUMemoryReporter::Delete(void* ptr) {
   }
   if (profile_memory) {
     reportMemoryUsageToProfiler(
-        ptr, -nbytes, c10::Device(c10::DeviceType::CPU));
+        ptr, nbytes, allocated, 0, c10::Device(c10::DeviceType::CPU));
```

Comment:
why did nbytes sign change?

Reply:
whoops, a mistake when bringing it back. fixed

Reply:
Hm, a mistake like this that's not caught by the tests is making me very nervous. Can you please add a test that would catch this?

Reply:
Tests are added for CPU and CUDA. Some minor bug are found during testing. 

I changed the cuda `DeviceCachingAllocator` to report Block wrapped ptr. It will be unwrapped by `THCCachingAllocator` and be wrapped it again as `DataPtr`. This will unify reported ptr semantic for CPU and CUDA allocator.

A bug is found due to `free_block` will change the passed in `Block`.

See the commit for details.